compiler/rustc_resolve/src/late.rs

```diff
@@ -300,6 +304,37 @@ impl LifetimeRib {
     }
 }
 
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
+struct MissingLifetime {
+    span: Span,
+    kind: MissingLifetimeKind,
+    count: usize,
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
+enum MissingLifetimeKind {
+    /// An explicit `'_`.
+    Underscore,
+    /// An elided lifetime `&' ty`.
+    Ampersand,
+    /// An elided lifetime in brackets with written brackets.
+    Comma,
+    /// An elided lifetime with elided brackets.
+    Brackets,
+}
+
+#[derive(Clone, Debug)]
+struct ElisionFailureInfo {
+    /// The index of the argument in the original definition.
+    index: usize,
+    /// The name of the argument if it's a simple ident.
+    ident: Option<Ident>,
+    /// The number of lifetimes in the parameter.
+    lifetime_count: usize,
+    /// The span of the parameter.
+    span: Span,
+}
```

Comment:
The `MissingLifetime`/`MissingLifetimeKind`/`ElisionFailureInfo` structures feel like a `diagnostics.rs` material.

Reply:
`LifetimeElisionCandidate` too.